### ISSUE TITLE
[v5] Stop conflating IDs with 'id' columns

### ISSUE THREAD

**Issue by Dreamsorcerer** [marmelab/react-admin#9516]
There are 2 concepts that react-admin is mixing up (I'll use these terms below to differentiate between them):

- ID: A unique ID that can be used to identify a record within react-admin and the API.
- 'id' column: A column/attribute of a row/record in the database/storage with the name 'id'.

This design decision causes all sorts of problems when trying to handle edge cases reliably. I would hugely appreciate, if there is a v5 in the future (I do see a milestone exists), that these 2 concepts are separated. Throughout the development of the aiohttp-admin library, these are some of the issues I've encountered that are caused by this (the ones I can recall today):

- A resource can't have an 'id' column if it's not the _sole_ primary key. It would be incredibly hacky to try and make this work, by internally renaming the column and remembering to change the label etc. anytime it's displayed. It's already hacky enough trying to get it to work with all the other issues, so I'm not even going to attempt this. The library currently just emits a warning and fails badly if a user provides such a model.
- An ID should be uniform across the library, so I use strings to represent IDs. This then means we need a bunch of hacks, as when an 'id' column is not a string, I need to remember to convert between string and the native type throughout the library at numerous locations. It has been very fiddly and required a lot of regression tests to ensure that we are getting the correct type at various points in the library.
- When a resource does _not_ have an 'id' column, we then again need a bunch of hacks to remember to remove the column from data at various points when submitted to the API by react-admin.
- Even with these hacks, there are still more edge cases that I keep finding where the behaviour is wrong. For example, the export button grabs the data from getMany() so it ends up exporting rows that include the ID. The ID is meant to be internal to react-admin and doesn't exist in the data, therefore it should not appear in an export. The API doesn't know if the call is for export or not, so it can't drop the ID there.

I'm sure you can come up with solutions, but as a 1 possible approach, a resource could be represented something like:
```
{
  'id': ...  // The ID of the react-admin resource.
  'data': {
    'id': ... // The 'id' column of the data, which may or may not be the same, and may not exist at all.
    ...
  }
}
```

If such a change were made, it would make react-admin a lot more reliable when trying to use it with arbitrary data models that haven't been pre-designed for react-admin.

**Comment by fzaninotto** [comment]
Thanks for your feedback.

Configurable ID for resources is indeed a common request, but it adds a tremendous amount of complexity everywhere in react-admin. We don't want this added complexity, and since there are workarounds, we don't want to add support for it. 

I invite you to fork react-admin and to try and add this configurability yourself. If you manage to do it without too much complexity, feel free to PR against the `next` branch (our current work on v5 is indeed the best time to consider this). But we won't do it ourselves.

**Comment by Dreamsorcerer** [comment]
@fzaninotto I realised that my proposal above might already be possible with very little change to react-admin, and I've spent the last couple of weeks trying it out. If there's interest in supporting this semi-officially, I'd be happy to write it up in a section of the documentation. However, there is one complete blocker I have currently, maybe something that could be tweaked in react-admin to make it easier?

The problem is getting ReferenceInput to work with this structure. Say we have 2 resources like this:
```
Author = {
  id: str
  data: {
    id: int  // PK in DB
  }
}
Book = {
  id: str
  data: {
    ...
    authorId: int  // FK in DB
  }
  fk_authorId: str
}
```

So, here author.data.id is the actual PK in the DB, while author.id is a normalised ID for the react-admin-compatible API.
Likewise book.data.authorId is the actual FK value in the DB, while fk_authorId is a normalised value that will match author.id.

What I'm trying to get ReferenceInput to do, is to use the normalised ID to do the lookup, but saving the data value.
So, roughly:

1. Get the author record with `getOne(book.fk_authorId)`.
2. Set `book.data.authorId = author.data.Id`.

Currently, I think both of these steps are handled by `source`. So, maybe a new prop could be added (that defaulted to using `source`) for one of these roles? Or some way of injecting a function which receives both record objects and can make the necessary changes after a selection has been made?

If I can get ReferenceInput working, then I believe I have a fully flexible solution which supports arbitrarily named PKs of arbitrary types, composite keys etc.
